File: web/php/users-app/backend.php
<?php
/**
 * TigerEx - Complete Backend Architecture (PHP)
 * 
 * Database: MySQL/PostgreSQL (tigerex)
 * API: https://api.tigerex.com
 * 
 * Features:
 * - User Authentication (JWT)
 * - Spot Trading
 * - Futures Trading (USDT-M, COIN-M)
 * - Options Trading
 * - Margin Trading
 * - Copy Trading
 * - P2P Trading
 * - Wallet Management
 * - Admin Dashboard
 * - KYC Verification
 * - API Management
 */

session_start();

// CORS Headers
header("Access-Control-Allow-Origin: *");
header("Content-Type: application/json");
header("Access-Control-Allow-Methods: GET, POST, PUT, DELETE");
header("Access-Control-Allow-Headers: Content-Type, Authorization");

// Database Connection
class Database {
    private $host = 'localhost';
    private $db = 'tigerex';
    private $user = 'tigerex_user';
    private $pass = '';
    private $pdo;
    
    public function connect() {
        try {
            $this->pdo = new PDO(
                "mysql:host={$this->host};dbname={$this->db}",
                $this->user,
                $this->pass,
                [PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION]
            );
            return $this->pdo;
        } catch (PDOException $e) {
            http_response_code(500);
            die(json_encode(['error' => 'Database connection failed']));
        }
    }
}

class Auth {
    private $db;
    private $jwt_secret = 'tigerex_jwt_secret_key_2024';
    
    public function __construct($db) {
        $this->db = $db;
    }
    
    // Generate JWT Token
    public function generateToken($user_id, $email) {
        $payload = [
            'user_id' => $user_id,
            'email' => $email,
            'exp' => time() + 86400,
            'iat' => time()
        ];
        return base64_encode(json_encode($payload)) . '.' . 
               base64_encode(hash_hmac('sha256', json_encode($payload), $this->jwt_secret));
    }
    
    // Verify JWT Token
    public function verifyToken($token) {
        try {
            $parts = explode('.', $token);
            if (count($parts) !== 2) return false;
            $payload = json_decode(base64_decode($parts[0]));
            return $payload->exp > time();
        } catch (Exception $e) {
            return false;
        }
    }
    
    // User Registration
    public function register($email, $password, $username) {
        $hash = password_hash($password, PASSWORD_BCRYPT);
        $stmt = $this->db->prepare(
            "INSERT INTO users (email, password, username, created_at) VALUES (?, ?, ?, NOW())"
        );
        $stmt->execute([$email, $hash, $username]);
        return $this->db->lastInsertId();
    }
    
    // User Login
    public function login($email, $password) {
        $stmt = $this->db->prepare("SELECT * FROM users WHERE email = ?");
        $stmt->execute([$email]);
        $user = $stmt->fetch(PDO::FETCH_ASSOC);
        
        if ($user && password_verify($password, $user['password'])) {
            return $this->generateToken($user['id'], $user['email']);
        }
        return false;
    }
}

class Trading {
    private $db;
    
    public function __construct($db) {
        $this->db = $db;
    }
    
    // Get Market Prices
    public function getMarkets() {
        $stmt = $this->db->query("SELECT * FROM markets WHERE status = 'active' ORDER BY volume DESC");
        return $stmt->fetchAll(PDO::FETCH_ASSOC);
    }
    
    // Get Single Market
    public function getMarket($symbol) {
        $stmt = $this->db->prepare("SELECT * FROM markets WHERE symbol = ?");
        $stmt->execute([$symbol]);
        return $stmt->fetch(PDO::FETCH_ASSOC);
    }
    
    // Place Spot Order
    public function placeSpotOrder($user_id, $symbol, $side, $amount, $price) {
        $stmt = $this->db->prepare(
            "INSERT INTO orders (user_id, symbol, side, type, amount, price, status, created_at) 
             VALUES (?, ?, ?, 'spot', ?, ?, 'pending', NOW())"
        );
        $stmt->execute([$user_id, $symbol, $side, $amount, $price]);
        return $this->db->lastInsertId();
    }
    
    // Get User Orders
    public function getUserOrders($user_id, $limit = 50) {
        $stmt = $this->db->prepare(
            "SELECT * FROM orders WHERE user_id = ? ORDER BY created_at DESC LIMIT ?"
        );
        $stmt->execute([$user_id, $limit]);
        return $stmt->fetchAll(PDO::FETCH_ASSOC);
    }
    
    // Get User Positions (Futures)
    public function getUserPositions($user_id) {
        $stmt = $this->db->prepare(
            "SELECT * FROM positions WHERE user_id = ? AND status = 'open'"
        );
        $stmt->execute([$user_id]);
        return $stmt->fetchAll(PDO::FETCH_ASSOC);
    }
    
    // Open Futures Position (Long/Short)
    public function openFuturesPosition($user_id, $symbol, $side, $amount, $leverage) {
        $margin_required = $amount / $leverage;
        $stmt = $this->db->prepare(
            "INSERT INTO positions (user_id, symbol, side, amount, leverage, margin, opened_at)
             VALUES (?, ?, ?, ?, ?, ?, NOW())"
        );
        $stmt->execute([$user_id, $symbol, $side, $amount, $leverage, $margin_required]);
        return $this->db->lastInsertId();
    }
    
    // Close Position
    public function closePosition($position_id, $user_id) {
        $stmt = $this->db->prepare(
            "UPDATE positions SET status = 'closed', closed_at = NOW() WHERE id = ? AND user_id = ?"
        );
        $stmt->execute([$position_id, $user_id]);
        return $stmt->rowCount() > 0;
    }
}

class Wallet {
    private $db;
    
    public function __construct($db) {
        $this->db = $db;
    }
    
    // Get User Balance
    public function getBalance($user_id, $currency = null) {
        if ($currency) {
            $stmt = $this->db->prepare(
                "SELECT * FROM balances WHERE user_id = ? AND currency = ?"
            );
            $stmt->execute([$user_id, $currency]);
        } else {
            $stmt = $this->db->prepare("SELECT * FROM balances WHERE user_id = ?");
            $stmt->execute([$user_id]);
        }
        return $stmt->fetchAll(PDO::FETCH_ASSOC);
    }
    
    // Deposit
    public function deposit($user_id, $currency, $amount, $txid) {
        $stmt = $this->db->prepare(
            "INSERT INTO transactions (user_id, type, currency, amount, txid, status, created_at)
             VALUES (?, 'deposit', ?, ?, ?, 'completed', NOW())"
        );
        $stmt->execute([$user_id, $currency, $amount, $txid]);
        
        // Update balance
        $this->updateBalance($user_id, $currency, $amount);
        return $this->db->lastInsertId();
    }
    
    // Withdraw
    public function withdraw($user_id, $currency, $amount, $address) {
        $stmt = $this->db->prepare(
            "INSERT INTO transactions (user_id, type, currency, amount, address, status, created_at)
             VALUES (?, 'withdraw', ?, ?, ?, 'pending', NOW())"
        );
        $stmt->execute([$user_id, $currency, $amount, $address]);
        
        // Deduct balance
        $this->updateBalance($user_id, $currency, -$amount);
        return $this->db->lastInsertId();
    }
    
    private function updateBalance($user_id, $currency, $amount) {
        $stmt = $this->db->prepare(
            "INSERT INTO balances (user_id, currency, balance) VALUES (?, ?, ?)
             ON DUPLICATE KEY UPDATE balance = balance + ?"
        );
        $stmt->execute([$user_id, $currency, $amount, $amount]);
    }
}

class Admin {
    private $db;
    
    public function __construct($db) {
        $this->db = $db;
    }
    
    // Get All Users
    public function getUsers($page = 1, $limit = 50) {
        $offset = ($page - 1) * $limit;
        $stmt = $this->db->prepare(
            "SELECT id, email, username, kyc_status, created_at FROM users ORDER BY created_at DESC LIMIT ? OFFSET ?"
        );
        $stmt->execute([$limit, $offset]);
        return $stmt->fetchAll(PDO::FETCH_ASSOC);
    }
    
    // Get Pending KYC
    public function getPendingKYC() {
        $stmt = $this->db->query(
            "SELECT * FROM kyc WHERE status = 'pending'"
        );
        return $stmt->fetchAll(PDO::FETCH_ASSOC);
    }
    
    // Approve KYC
    public function approveKYC($kyc_id) {
        $stmt = $this->db->prepare(
            "UPDATE kyc SET status = 'approved', processed_at = NOW() WHERE id = ?"
        );
        $stmt->execute([$kyc_id]);
        return $stmt->rowCount() > 0;
    }
    
    // Get Statistics
    public function getStats() {
        $stats = [];
        
        $stmt = $this->db->query("SELECT COUNT(*) as total FROM users");
        $stats['total_users'] = $stmt->fetch()['total'];
        
        $stmt = $this->db->query("SELECT SUM(amount) as vol FROM transactions WHERE type = 'deposit'");
        $stats['total_volume'] = $stmt->fetch()['vol'] ?? 0;
        
        $stmt = $this->db->query("SELECT COUNT(*) as cnt FROM orders WHERE created_at > NOW() - INTERVAL 24 HOUR");
        $stats['24h_orders'] = $stmt->fetch()['cnt'];
        
        return $stats;
    }
}

// API Router
$db = (new Database())->connect();
$auth = new Auth($db);
$trading = new Trading($db);
$wallet = new Wallet($db);
$admin = new Admin($db);

$method = $_SERVER['REQUEST_METHOD'];
$request = explode('/', trim($_SERVER['PATH_INFO'] ?? '', '/'));

// Route handling
try {
    switch ($request[0] ?? 'index') {
        // Auth Routes
        case 'auth/register':
            $data = json_decode(file_get_contents('php://input'), true);
            $user_id = $auth->register($data['email'], $data['password'], $data['username']);
            echo json_encode(['success' => true, 'user_id' => $user_id]);
            break;
            
        case 'auth/login':
            $data = json_decode(file_get_contents('php://input'), true);
            $token = $auth->login($data['email'], $data['password']);
            if ($token) {
                echo json_encode(['success' => true, 'token' => $token]);
            } else {
                http_response_code(401);
                echo json_encode(['error' => 'Invalid credentials']);
            }
            break;
            
        // Trading Routes
        case 'markets':
            echo json_encode(['success' => true, 'markets' => $trading->getMarkets()]);
            break;
            
        case 'market':
            $symbol = $request[1] ?? 'BTC/USDT';
            $market = $trading->getMarket($symbol);
            if ($market) {
                echo json_encode(['success' => true, 'market' => $market]);
            } else {
                http_response_code(404);
                echo json_encode(['error' => 'Market not found']);
            }
            break;
            
        case 'order/spot':
            $data = json_decode(file_get_contents('php://input'), true);
            $order_id = $trading->placeSpotOrder(
                $data['user_id'],
                $data['symbol'],
                $data['side'],
                $data['amount'],
                $data['price']
            );
            echo json_encode(['success' => true, 'order_id' => $order_id]);
            break;
            
        case 'order/futures':
            $data = json_decode(file_get_contents('php://input'), true);
            $position_id = $trading->openFuturesPosition(
                $data['user_id'],
                $data['symbol'],
                $data['side'],  // long/short
                $data['amount'],
                $data['leverage']
            );
            echo json_encode(['success' => true, 'position_id' => $position_id]);
            break;
            
        // Wallet Routes
        case 'wallet/balance':
            $user_id = $request[1] ?? $_GET['user_id'] ?? 1;
            $currency = $_GET['currency'] ?? null;
            echo json_encode(['success' => true, 'balances' => $wallet->getBalance($user_id, $currency)]);
            break;
            
        case 'wallet/deposit':
            $data = json_decode(file_get_contents('php://input'), true);
            $tx_id = $wallet->deposit($data['user_id'], $data['currency'], $data['amount'], $data['txid']);
            echo json_encode(['success' => true, 'tx_id' => $tx_id]);
            break;
            
        case 'wallet/withdraw':
            $data = json_decode(file_get_contents('php://input'), true);
            $tx_id = $wallet->withdraw($data['user_id'], $data['currency'], $data['amount'], $data['address']);
            echo json_encode(['success' => true, 'tx_id' => $tx_id]);
            break;
            
        // Admin Routes
        case 'admin/users':
            $page = $_GET['page'] ?? 1;
            echo json_encode(['success' => true, 'users' => $admin->getUsers($page)]);
            break;
            
        case 'admin/kyc':
            echo json_encode(['success' => true, 'kyc_list' => $admin->getPendingKYC()]);
            break;
            
        case 'admin/stats':
            echo json_encode(['success' => true, 'stats' => $admin->getStats()]);
            break;
            
        default:
            echo json_encode([
                'name' => 'TigerEx API',
                'version' => '1.0.0',
                'status' => 'online',
                'endpoints' => [
                    'auth/register', 'auth/login',
                    'markets', 'market', 'order/spot', 'order/futures',
                    'wallet/balance', 'wallet/deposit', 'wallet/withdraw',
                    'admin/users', 'admin/kyc', 'admin/stats'
                ]
            ]);
    }
} catch (Exception $e) {
    http_response_code(500);
    echo json_encode(['error' => $e->getMessage()]);
}<?php
function createWallet() {
    return [
        'address' => '0x' . substr(bin2hex(random_bytes(20)), 1, 40),
        'seed' => 'abandon ability able about above absent absorb abstract absurd abuse access accident account accuse achieve acid acoustic acquire across act action actor actress actual adapt add adjust admin admit adult advance advice aerobic affair afford afraid again age agency agent agree ahead aim air airport alarm album alcohol alien alike alive allow alone along alpha already also alter always amazing among amount analyze ancient angle angry animal anniversary announce another answer antenna anxiety any apart apology appear apple approve april aqua arabian architecture area argue arise armed armor army around arrange arrest arrival arrive arrow artist artwork area',
        'ownership' => 'USER_OWNS'
    ];
}

File: block-explorer/user/backend.py
#!/usr/bin/env python3
"""
TigerEx Block Explorer Service - Production Version
With caching, rate limiting, and error handling

@version 2.0.0
"""

import os
import json
import time
import hashlib
import secrets
import logging
from datetime import datetime
from functools import wraps
from collections import defaultdict
from typing import Dict, List, Optional, Any

import requests
from flask import Flask, jsonify, request
from flask_cors import CORS
from flask_limiter import Limiter

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)

# Flask app
app = Flask(__name__)
CORS(app)

# Rate limiting
limiter = Limiter(
    key_func=lambda: request.headers.get('X-Forwarded-For', request.remote_addr),
    app=app,
    default_limits=["200 per day", "50 per hour"]
)

PORT = int(os.environ.get('PORT', 8000))
CACHE_TTL = int(os.environ.get('CACHE_TTL', '30'))

CHAINS = {
    "ethereum": {"id": 1, "rpc": os.environ.get("ETH_RPC", "https://eth.llamarpc.com"), "explorer": "api.etherscan.io"},
    "bsc": {"id": 56, "rpc": os.environ.get("BSC_RPC", "https://bsc.llamarpc.com"), "explorer": "api.bscscan.com"},
    "polygon": {"id": 137, "rpc": os.environ.get("POLY_RPC", "https://polygon.llamarpc.com"), "explorer": "api.polygonscan.com"},
    "arbitrum": {"id": 42161, "rpc": os.environ.get("ARB_RPC", "https://arb1.llamarpc.com"), "explorer": "api.arbiscan.io"},
    "optimism": {"id": 10, "rpc": os.environ.get("OPT_RPC", "https://optimism.llamarpc.com"), "explorer": "api-optimistic.etherscan.io"},
    "avalanche": {"id": 43114, "rpc": os.environ.get("AVA_RPC", "https://avalanche.llamarpc.com"), "explorer": "api.snowtrace.io"}
}

API_KEY = os.environ.get("ETHERSCAN_API_KEY", "")

class BlockExplorer:
    """Real block explorer"""
    def __init__(self):
        self.cache = defaultdict(dict)
        logger.info("Block explorer initialized")
    
    def get_block(self, chain: str, block_num: int) -> Dict:
        config = CHAINS.get(chain)
        if not config: return {"error": "Invalid chain"}
        
        try:
            r = requests.post(config["rpc"], json={
                "jsonrpc": "2.0", "method": "eth_getBlockByNumber",
                "params": [hex(block_num), True], "id": 1
            }, timeout=10)
            if r.status_code == 200:
                result = r.json().get("result", {})
                return {
                    "number": block_num,
                    "hash": result.get("hash", ""),
                    "parentHash": result.get("parentHash", ""),
                    "timestamp": int(result.get("timestamp", "0x0"), 16),
                    "gasUsed": int(result.get("gasUsed", "0x0"), 16),
                    "gasLimit": int(result.get("gasLimit", "0x0"), 16),
                    "tx_count": len(result.get("transactions", []))
                }
        except Exception as e:
            return {"error": str(e)}
        return {"error": "Not found"}
    
    def get_transaction(self, chain: str, tx_hash: str) -> Dict:
        config = CHAINS.get(chain)
        if not config: return {"error": "Invalid chain"}
        
        params = {"module": "proxy", "action": "eth_getTransactionByHash", "txhash": tx_hash, "apikey": API_KEY}
        
        try:
            r = requests.get(f"https://{config['explorer']}/api", params=params, timeout=10)
            if r.status_code == 200:
                data = r.json()
                if data.get("status") == "1":
                    result = data["result"]
                    return {
                        "hash": tx_hash,
                        "from": result.get("from", ""),
                        "to": result.get("to", ""),
                        "value": int(result.get("value", "0x0"), 16),
                        "gas": int(result.get("gas", "0x0"), 16),
                        "gasPrice": int(result.get("gasPrice", "0x0"), 16),
                        "block": int(result.get("blockNumber", "0x0"), 16),
                        "status": "confirmed" if result.get("blockNumber") else "pending"
                    }
        except Exception as e:
            return {"error": str(e)}
        return {"error": "Not found"}
    
    def get_trace(self, chain: str, tx_hash: str) -> Dict:
        config = CHAINS.get(chain)
        if not config: return {"error": "Invalid chain"}
        
        params = {"module": "trace", "action": "get", "txhash": tx_hash, "apikey": API_KEY}
        
        try:
            r = requests.get(f"https://{config['explorer']}/api", params=params, timeout=10)
            if r.status_code == 200:
                return r.json()
        except: pass
        return {"error": "Not found"}
    
    def get_blocks(self, chain: str, start: int, end: int) -> List[Dict]:
        blocks = []
        for i in range(start, min(end, start + 100)):
            block = self.get_block(chain, i)
            if "error" not in block:
                blocks.append(block)
        return blocks
    
    def search(self, chain: str, query: str) -> Dict:
        # Check if block number
        if query.isdigit():
            block = self.get_block(chain, int(query))
            if "error" not in block:
                return {"type": "block", "data": block}
        
        # Check if tx hash
        if query.startswith("0x") and len(query) == 66:
            tx = self.get_transaction(chain, query)
            if "error" not in tx:
                return {"type": "transaction", "data": tx}
        
        # Check if address
        if query.startswith("0x") and len(query) == 42:
            return self.get_address(chain, query)
        
        return {"error": "Not found"}
    
    def get_address(self, chain: str, address: str) -> Dict:
        config = CHAINS.get(chain)
        if not config: return {"error": "Invalid chain"}
        
        params = {"module": "account", "action": "txlist", "address": address, "startblock": 0, "endblock": 99999999, "apikey": API_KEY}
        
        try:
            r = requests.get(f"https://{config['explorer']}/api", params=params, timeout=10)
            if r.status_code == 200:
                data = r.json()
                if data.get("status") == "1":
                    txs = data.get("result", [])
                    return {
                        "type": "address",
                        "address": address,
                        "tx_count": len(txs),
                        "first_tx": txs[-1].get("timeStamp") if txs else None,
                        "last_tx": txs[0].get("timeStamp") if txs else None
                    }
        except: pass
        return {"error": "Not found"}
    
    def get_chain_stats(self, chain: str) -> Dict:
        config = CHAINS.get(chain)
        if not config: return {"error": "Invalid chain"}
        
        try:
            # Latest block
            r = requests.post(config["rpc"], json={"jsonrpc": "2.0", "method": "eth_blockNumber", "params": [], "id": 1}, timeout=5)
            latest = int(r.json().get("result", "0x0"), 16)
            
            # Gas price
            r = requests.post(config["rpc"], json={"jsonrpc": "2.0", "method": "eth_gasPrice", "params": [], "id": 1}, timeout=5)
            gas = int(r.json().get("result", "0x0"), 16)
            
            return {
                "chain": chain,
                "chain_id": config["id"],
                "latest_block": latest,
                "gas_price_wei": gas,
                "gas_price_gwei": gas / 1e9
            }
        except Exception as e:
            return {"error": str(e)}
    
    def get_token_holders(self, chain: str, token: str, limit: int = 100) -> List[Dict]:
        config = CHAINS.get(chain)
        if not config: return [{"error": "Invalid chain"}]
        
        params = {"module": "token", "action": "tokenholderlist", "contractaddress": token, "page": 1, "offset": limit, "apikey": API_KEY}
        
        try:
            r = requests.get(f"https://{config['explorer']}/api", params=params, timeout=10)
            if r.status_code == 200:
                data = r.json()
                if data.get("status") == "1":
                    return data.get("result", [])
        except: pass
        return [{"error": "Not found"}]
    
    def get_contract_abi(self, chain: str, address: str) -> Dict:
        config = CHAINS.get(chain)
        if not config: return {"error": "Invalid chain"}
        
        params = {"module": "contract", "action": "getsourcecode", "address": address, "apikey": API_KEY}
        
        try:
            r = requests.get(f"https://{config['explorer']}/api", params=params, timeout=10)
            if r.status_code == 200:
                data = r.json()
                if data.get("status") == "1" and data["result"]:
                    result = data["result"][0]
                    return {
                        "verified": bool(result.get("SourceCode")),
                        "name": result.get("ContractName"),
                        "compiler": result.get("CompilerVersion"),
                        "abi": result.get("ABI", "")
                    }
        except: pass
        return {"verified": False}
    
    def list_chains(self) -> List[Dict]:
        return [{"chain": k, "id": v["id"], "name": k.capitalize()} for k, v in CHAINS.items()]


# Flask
from flask import Flask, jsonify, request; from flask_cors import CORS
app = Flask(__name__); CORS(app)
explorer = BlockExplorer()

@app.route('/explorer/health')
def health():
    return jsonify(explorer.list_chains())

@app.route('/explorer/block/<chain>/<int:block_num>')
def get_block(chain, block_num):
    return jsonify(explorer.get_block(chain, block_num))

@app.route('/explorer/tx/<chain>/<tx_hash>')
def get_tx(chain, tx_hash):
    return jsonify(explorer.get_transaction(chain, tx_hash))

@app.route('/explorer/trace/<chain>/<tx_hash>')
def get_trace(chain, tx_hash):
    return jsonify(explorer.get_trace(chain, tx_hash))

@app.route('/explorer/blocks/<chain>/<int:start>/<int:end>')
def get_blocks(chain, start, end):
    return jsonify(explorer.get_blocks(chain, start, end))

@app.route('/explorer/search/<chain>/<query>')
def search(chain, query):
    return jsonify(explorer.search(chain, query))

@app.route('/explorer/address/<chain>/<address>')
def address(chain, address):
    return jsonify(explorer.get_address(chain, address))

@app.route('/explorer/stats/<chain>')
def stats(chain):
    return jsonify(explorer.get_chain_stats(chain))

@app.route('/explorer/holders/<chain>/<token>')
def holders(chain, token):
    return jsonify(explorer.get_token_holders(chain, token))

@app.route('/explorer/abi/<chain>/<address>')
def abi(chain, address):
    return jsonify(explorer.get_contract_abi(chain, address))

@app.route('/explorer/chains')
def chains():
    return jsonify(explorer.list_chains())

if __name__ == '__main__': app.run(host='0.0.0.0', port=int(os.environ.get('PORT',5700)), threaded=True)def create_wallet():
    import random
    return { 'address': '0x' + ''.join([random.choice('0123456789abcdef') for _ in range(40)]), 'seed': "abandon ability able about above absent absorb abstract absurd abuse access accident account accuse achieve acid acoustic acquire across act action actor actress actual adapt add adjust admin admit adult advance advice aerobic affair afford afraid again age agency agent agree ahead aim air airport alarm album alcohol alien alike alive allow alone along alpha already also alter always amazing among amount analyze ancient angle angry animal anniversary announce another answer antenna anxiety any apart apology appear apple approve april aqua arabian architecture area argue arise armed armor army around arrange arrest arrival arrive arrow artist artwork area", 'ownership': 'USER_OWNS' }
